src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -282,6 +283,36 @@ impl KvmVcpu {
         Ok(cpuid)
     }
 
+    /// If the IA32_TSC_DEADLINE MSR value is zero, update it
+    /// with the IA32_TSC value to guarantee that
+    /// the vCPU will continue receiving interrupts after restoring from a snapshot.
+    ///
+    /// Rationale: we observed that sometimes when taking a snapshot,
+    /// the IA32_TSC_DEADLINE MSR is cleared, but the interrupt is not
+    /// delivered to the guest, leading to a situation where one
+    /// of the vCPUs never receives TSC interrupts after restoring,
+    /// until the MSR is updated externally, eg by setting the system time.
+    fn fix_zero_tsc_deadline_msr(msr_chunks: &mut [Msrs]) {
+        for msrs in msr_chunks.iter_mut() {
+            let tsc_value = msrs
+                .as_slice()
+                .iter()
+                .find(|msr| msr.index == MSR_IA32_TSC)
+                .map(|msr| msr.data)
+                .unwrap_or(0);
+
+            for msr in msrs.as_mut_slice().iter_mut() {
+                if msr.index == MSR_IA32_TSC_DEADLINE && msr.data == 0 {
+                    warn!(
+                        "MSR_IA32_TSC_DEADLINE is 0, replacing with {:x}.",
+                        tsc_value
+                    );
+                    msr.data = tsc_value;
+                }
+            }
```

Comment:
As I understand both `TSC` and `TSC_DEADLINE` need to be present in the msr list for this to work right?
The ` .unwrap_or(0)` line is a bit confusing. If `TSC_DEADLINE` is 0 then there is no point of setting it to 0 again.
Think this version will be more clear:
```Rust
// If no TSC found we cannot do anything
let Some(tsc_msr) = msrs
                .as_slice()
                .iter()
                .find(|msr| msr.index == MSR_IA32_TSC).cloned() else { continue; };
// If no tsc_deadline we cannot do anything
let Some(tsc_deadline_msr) = msrs
                .as_mut_slice()
                .iter_mut()
                .find(|msr| msr.index == MSR_IA32_TSC) else { continue; };
if tsc_deadline_msr.data == 0 {
  warn!(
    "MSR_IA32_TSC_DEADLINE is 0, replacing with {:x}.",
    tsc_value
  );
  tsc_deadline_msr.data = tsc_msr.data;
}
```

Reply:
While I was thinking about this, I realised that the original code was not properly handling the case where TSC and TSC_DEADLINE MSRs were in separate chunks. I fixed that part and added a test.